File: app/src/main/java/com/example/myapplication11/base/BaseFragment.java
package com.example.myapplication11.base;

import android.app.Activity;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.databinding.DataBindingUtil;
import androidx.databinding.ViewDataBinding;
import androidx.fragment.app.Fragment;
import androidx.lifecycle.Observer;

import com.example.myapplication11.R;
import com.example.myapplication11.config.LoadState;
import com.example.myapplication11.databinding.FragmentBaseBinding;
import com.example.myapplication11.databinding.ViewLoadErrorBinding;
import com.example.myapplication11.databinding.ViewLoadingBinding;
import com.example.myapplication11.databinding.ViewNoDataBinding;
import com.example.myapplication11.databinding.ViewNoNetworkBinding;

import com.example.myapplication11.manager.MyActivityManager;
import com.example.myapplication11.ui.activity.login.LoginActivity;
import com.example.myapplication11.ui.activity.main.MainActivity;
import com.example.myapplication11.ui.activity.main.MainViewModel;

public abstract class BaseFragment<DB extends ViewDataBinding, VM extends BaseViewModel> extends Fragment {
    protected DB mDataBinding;
    protected VM mViewModel;

    private FragmentBaseBinding mFragmentBaseBinding;
    private ViewLoadingBinding mViewLoadingBinding;
    private ViewLoadErrorBinding mViewLoadErrorBinding;
    private ViewNoNetworkBinding mViewNoNetworkBinding;
    private ViewNoDataBinding mViewNoDataBinding;

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        Bundle args = getArguments();
        if (args != null) {
            handleArguments(args);
        }
        initViewModel();
        if (mViewModel != null) {
            getLifecycle().addObserver(mViewModel);
        }
    }

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        mFragmentBaseBinding = DataBindingUtil.inflate(inflater, R.layout.fragment_base, container, false);
        mDataBinding = DataBindingUtil.inflate(inflater,
                getLayoutResId(), mFragmentBaseBinding.flContentContainer, true);

        bindViewModel();
        mDataBinding.setLifecycleOwner(this);
        initLoadState();
        initCollectState();
        init();
        return mFragmentBaseBinding.getRoot();
    }

    private void initLoadState() {
        if (mViewModel != null && isSupportLoad()) {
            mViewModel.loadState.observe(getViewLifecycleOwner(), new Observer<LoadState>() {
                @Override
                public void onChanged(LoadState loadState) {
                    switchLoadView(loadState);
                }
            });
            Activity activity = MyActivityManager.getsInstance().getCurrentActivity();
            if (activity instanceof MainActivity) {
                ((MainActivity) activity).mDataBinding.fabTop.setVisibility(View.VISIBLE);
            }
        } else {
            Activity activity = MyActivityManager.getsInstance().getCurrentActivity();
            if (activity instanceof MainActivity) {
                ((MainActivity) activity).mDataBinding.fabTop.setVisibility(View.GONE);
            }
        }
    }


    private void switchLoadView(LoadState loadState) {
        removeLoadView();

        switch (loadState) {
            case LOADING:
                if (mViewLoadingBinding == null) {
                    mViewLoadingBinding = DataBindingUtil.inflate(getLayoutInflater(), R.layout.view_loading,
                            mFragmentBaseBinding.flContentContainer, false);
                }
                mFragmentBaseBinding.flContentContainer.addView(mViewLoadingBinding.getRoot());
                break;

            case NO_NETWORK:
                if (mViewNoNetworkBinding == null) {
                    mViewNoNetworkBinding = DataBindingUtil.inflate(getLayoutInflater(), R.layout.view_no_network,
                            mFragmentBaseBinding.flContentContainer, false);
                    mViewNoNetworkBinding.setViewModel(mViewModel);
                }
                mFragmentBaseBinding.flContentContainer.addView(mViewNoNetworkBinding.getRoot());
                break;

            case NO_DATA:
                if (mViewNoDataBinding == null) {
                    mViewNoDataBinding = DataBindingUtil.inflate(getLayoutInflater(), R.layout.view_no_data,
                            mFragmentBaseBinding.flContentContainer, false);
                }
                mFragmentBaseBinding.flContentContainer.addView(mViewNoDataBinding.getRoot());
                break;

            case ERROR:
                if (mViewLoadErrorBinding == null) {
                    mViewLoadErrorBinding = DataBindingUtil.inflate(getLayoutInflater(), R.layout.view_load_error,
                            mFragmentBaseBinding.flContentContainer, false);
                    mViewLoadErrorBinding.setViewModel(mViewModel);
                }
                mFragmentBaseBinding.flContentContainer.addView(mViewLoadErrorBinding.getRoot());
                break;

            default:
                break;
        }
    }

    private void removeLoadView() {
        int childCount = mFragmentBaseBinding.flContentContainer.getChildCount();
        if (childCount > 1) {
            mFragmentBaseBinding.flContentContainer.removeViews(1, childCount - 1);
        }
    }


    private void initCollectState() {
        if (mViewModel == null) {
            return;
        }
        mViewModel.getCollect().observe(getViewLifecycleOwner(), new Observer<Object>() {
            @Override
            public void onChanged(Object collect) {
                if (collect == null) {
                    //跳转到登录界面
//                    LoginActivity.start(getActivity());
                }
            }
        });

    }

    @Override
    public void onDestroyView() {
        super.onDestroyView();
        if (mViewModel != null) {
            getLifecycle().removeObserver(mViewModel);
        }
        removeLoadView();
    }


    /**
     * 处理参数
     *
     * @param args 参数容器
     */
    protected void handleArguments(Bundle args) {

    }


    /**
     * 是否支持页面加载。默认不支持
     *
     * @return true表示支持，false表示不支持
     */
    protected boolean isSupportLoad() {
        return false;
    }

    /**
     * 获取当前页面的布局资源ID
     *
     * @return 布局资源ID
     */
    protected abstract int getLayoutResId();

    /**
     * 初始化ViewModel
     */
    protected abstract void initViewModel();

    /**
     * 绑定ViewModel
     */
    protected abstract void bindViewModel();

    /**
     * 初始化
     */
    protected abstract void init();

}

File: appkotlin/src/main/java/com/example/appkotlin/base/BaseFragment.kt
package com.example.appkotlin.base

import android.os.Bundle
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.core.view.children
import androidx.databinding.DataBindingUtil
import androidx.databinding.ViewDataBinding
import androidx.fragment.app.Fragment
import androidx.fragment.app.viewModels
import androidx.lifecycle.Observer
import androidx.lifecycle.ViewModel
import com.example.appkotlin.LoadState
import com.example.appkotlin.R
import com.example.appkotlin.databinding.*


abstract class BaseFragment<DM : ViewDataBinding, VM : BaseViewModel> : Fragment() {

    lateinit var mDataBinding: DM
    lateinit var mViewModel: VM

    lateinit var baseBinding: FragmentBaseBinding
    lateinit var noDataBinding: ItemNodataBinding
    lateinit var noNetBinding: ItemNonetBinding
    lateinit var loadingBinding: ItemLoadingBinding
    lateinit var loadErrorBinding: ItemErrorBinding

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        initViewModel()
        this.lifecycle.addObserver(mViewModel)
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?,
                              savedInstanceState: Bundle?): View? {
        // Inflate the layout for this fragment

        baseBinding = DataBindingUtil.inflate(inflater, R.layout.fragment_base, container, false)

        mDataBinding = DataBindingUtil.inflate(inflater, getResLayoutId(), baseBinding.flContent, true)
        bindViewModel()
        mDataBinding.lifecycleOwner = this
        initLoadState()

        init()
        return baseBinding.root
    }


    private fun initLoadState() {
        if (mViewModel != null) {
//            mViewModel.loadState.observe(viewLifecycleOwner, Observer<LoadState> { t: LoadState? ->
//                if (t != null) {
//                    Log.e("LoadState--->B1", "    " + t)
//                    switchLoadState(t)
//                }B
//            })
            mViewModel.loadState.observe(viewLifecycleOwner, Observer<LoadState>{
                Log.e("LoadState--->B1-",""+it);
            })
        }
        Log.e("LoadState--->B",""+"Observer");


    }


    private fun switchLoadState(loadState: LoadState) {

        removeLoading()

        Log.e("LoadState--->B", "    " + loadState)
        when (loadState) {
            LoadState.LOADING -> {
                if (!::loadingBinding.isInitialized || loadingBinding == null) {
                    loadingBinding = DataBindingUtil.inflate(layoutInflater, R.layout.item_loading, baseBinding.flContent, false);
                }
                baseBinding.flContent.addView(loadingBinding.root)
            }

            LoadState.NO_DATA -> {
                if (!::noDataBinding.isInitialized || noDataBinding == null) {
                    noDataBinding = DataBindingUtil.inflate(layoutInflater, R.layout.item_nodata, baseBinding.flContent, false)
                }
                baseBinding.flContent.addView(noDataBinding.root)
            }
            LoadState.NO_NETWORK -> {
                if (!::noNetBinding.isInitialized || noNetBinding == null) {
                    noNetBinding = DataBindingUtil.inflate(layoutInflater, R.layout.item_nonet, baseBinding.flContent, false)
                }
                baseBinding.flContent.addView(noNetBinding.root)
            }
            LoadState.LOAD_ERROR -> {
                if (!::loadErrorBinding.isInitialized || loadErrorBinding == null) {
                    loadErrorBinding = DataBindingUtil.inflate(layoutInflater, R.layout.item_error, baseBinding.flContent, false)
                }
                baseBinding.flContent.addView(loadErrorBinding.root)
            }
        }
    }

    fun removeLoading() {
        var childCount = baseBinding.flContent.childCount
        if (childCount > 1) {
            baseBinding.flContent.removeViews(1, childCount - 1)
        }
    }


    //获取当前  布局ID
    abstract fun getResLayoutId(): Int


    //实例化ViewModel
    abstract fun initViewModel()


    //绑定viewModel
    abstract fun bindViewModel()

    abstract fun init()

//    companion object {
//        /**
//         * Use this factory method to create a new instance of
//         * this fragment using the provided parameters.
//         *
//         * @param param1 Parameter 1.
//         * @param param2 Parameter 2.
//         * @return A new instance of fragment BaseFragment.
//         */
//        // TODO: Rename and change types and number of parameters
//        @JvmStatic
//        fun newInstance(param1: String, param2: String) =
//                BaseFragment().apply {
//                    arguments = Bundle().apply {
//                        putString(ARG_PARAM1, param1)
//                        putString(ARG_PARAM2, param2)
//                    }
//                }
//    }
}


